[GTK][WPE] Make it possible to verify damage rects in layout tests

spenap opened the pull request:
#### 74803a0710d915fb9d76175f7c22186f8315e54e
<pre>
[GTK][WPE] Make it possible to verify damage rects in layout tests
<a href="https://bugs.webkit.org/show_bug.cgi?id=284951">https://bugs.webkit.org/show_bug.cgi?id=284951</a>

Reviewed by Carlos Garcia Campos.

Expose damage information to JS through Internals for validation using
layout tests.

When DAMAGE_TRACKING is enabled and the Internals object is available,
frame damage information is stored and then made available through a new
FrameDamageForTesting interface. This is exposed through the
DrawingArea, and available in AC mode.

* LayoutTests/platform/glib/damage/basic-propagation-001-expected.txt: Added.
* LayoutTests/platform/glib/damage/basic-propagation-001.html: Added.
* LayoutTests/platform/glib/damage/basic-propagation-002-expected.txt: Added.
* LayoutTests/platform/glib/damage/basic-propagation-002.html: Added.
* LayoutTests/platform/glib/damage/common.css: Added.
(body):
* LayoutTests/platform/glib/damage/common.js: Added.
(failTest):
(setupTestCase):
* Source/WebCore/page/ChromeClient.h: New API to reset and get damage
  history.
(WebCore::ChromeClient::resetDamageHistoryForTesting):
(WebCore::ChromeClient::damageHistoryForTesting const):
* Source/WebCore/platform/graphics/Damage.h: New wrapper to store frame
  damage history.
(WebCore::FrameDamageHistory::damageInformation const):
(WebCore::FrameDamageHistory::addDamage):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::resetToConsistentState):
(WebCore::Internals::Internals):
(WebCore::Internals::getCurrentDamagePropagation const):
(WebCore::Internals::getFrameDamageHistory const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::resetDamageHistoryForTesting):
(WebKit::WebChromeClient::damageHistoryForTesting const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/DrawingAreaCoordinatedGraphics.cpp:
(WebKit::DrawingAreaCoordinatedGraphics::frameDamageForTesting const):
If AC mode is not enabled, frame damage history won&apos;t be available.
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/DrawingAreaCoordinatedGraphics.h:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/FrameDamageForTesting.h: Added.
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.h:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
(WebKit::ThreadedCompositor::paintToCurrentGLContext):
(WebKit::ThreadedCompositor::resetFrameDamageHistory):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.h:
(WebKit::ThreadedCompositor::frameDamageHistory const):
* Source/WebKit/WebProcess/WebPage/DrawingArea.h: Allow querying frame
  damage history for testing without side effects.
(WebKit::DrawingArea::frameDamageForTesting const):

Canonical link: <a href="https://commits.webkit.org/290948@main">https://commits.webkit.org/290948@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/647a8329e44a6d06557460d7cd606044c7e38351

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91538 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11069 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/600 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96507 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42224 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93588 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11447 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19485 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70295 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27812 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94539 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8740 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82934 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50619 "Found 1 new API test failure: /WPE/TestSSL:/webkit/WebKitWebView/ephemeral-tls-errors (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8503 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/523 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41394 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78823 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/526 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98510 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18699 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13776 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79317 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18954 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78772 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78519 "Found 1 new API test failure: /WebKitGTK/TestInspector:/webkit/WebKitWebInspector/manual-attach-detach (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19420 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23045 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/402 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/11824 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18695 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23971 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18406 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21866 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20172 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Scony commented on LayoutTests/damage/basic-propagation-001.html:
```diff
@@ -0,0 +1,33 @@
+
+      processAnimationFrameSequence([
+          () => {}, // Skip one RAF callback to make sure initial painting is done.
+          () => {
```
we need to assert if the damage propagation == Region as well

spenap replied:
`setupTestCase` does that already, so the test would have failed if `false`. But if you want to avoid `processAnimationFrameSequence` running, we could make `setupTestCase` return a boolean or throw an exception, and make this conditional to that

Scony replied:
good point - let's leave it as-is then

Scony commented on LayoutTests/damage/basic-propagation-002.html:
```diff
@@ -0,0 +1,55 @@
+
+      processAnimationFrameSequence([
+          () => {}, // Skip one RAF callback to make sure initial painting is done.
+          () => {
```
we need to assert if the damage propagation == Region as well

Scony commented on Source/WebCore/testing/Internals.h:
```diff
@@ -1546,6 +1550,18 @@ class Internals final
 
+#if ENABLE(DAMAGE_TRACKING)
+    using DamagePropagation = Damage::Propagation;
+    struct DamageDetails {
```
as this describes the damage of particular frame, I'd call it something more like `FrameDamage`

Scony commented on Source/WebCore/testing/Internals.h:
```diff
@@ -1546,6 +1550,18 @@ class Internals final
+        Vector<Ref<DOMRectReadOnly>> rects;
+    };
+    std::optional<DamagePropagation> getCurrentDamagePropagation() const;
+    ExceptionOr<Vector<DamageDetails>> getDamageDetails() const;
```
and here I'd lean towards something like `getFrameDamageHistory` or so

Scony commented on Source/WebCore/testing/Internals.h:
```diff
@@ -59,6 +59,10 @@
 #include "AudioSession.h"
 #endif
 
+#if ENABLE(DAMAGE_TRACKING)
+#include "Damage.h"
+#endif
```
`Damage.h` has a `#if ENABLE(DAMAGE_TRACKING)` guard internally so no need to guard here

Scony commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -1982,4 +1987,23 @@ void WebChromeClient::didProgrammaticallyClearTextFormControl(const HTMLTextForm
+#if ENABLE(DAMAGE_TRACKING)
+void WebChromeClient::resetDamageForTesting()
+{
+    auto layerTreeHost = static_cast<LayerTreeHost*>(graphicsLayerFactory());
```
```suggestion
    auto* layerTreeHost = static_cast<LayerTreeHost*>(graphicsLayerFactory());
```

Scony commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -1982,4 +1987,23 @@ void WebChromeClient::didProgrammaticallyClearTextFormControl(const HTMLTextForm
+{
+    auto layerTreeHost = static_cast<LayerTreeHost*>(graphicsLayerFactory());
+
+    auto threadedCompositor = layerTreeHost->threadedCompositor();
```
```suggestion
    auto* threadedCompositor = layerTreeHost->threadedCompositor();
```

Scony commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -1982,4 +1987,23 @@ void WebChromeClient::didProgrammaticallyClearTextFormControl(const HTMLTextForm
+
+WebCore::DamageForTesting* WebChromeClient::getDamageForTesting() const
+{
+    auto layerTreeHost = static_cast<LayerTreeHost*>(graphicsLayerFactory());
```
```suggestion
    auto* layerTreeHost = static_cast<LayerTreeHost*>(graphicsLayerFactory());
```

Scony commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -1982,4 +1987,23 @@ void WebChromeClient::didProgrammaticallyClearTextFormControl(const HTMLTextForm
+WebCore::DamageForTesting* WebChromeClient::getDamageForTesting() const
+{
+    auto layerTreeHost = static_cast<LayerTreeHost*>(graphicsLayerFactory());
+    const auto threadedCompositor = layerTreeHost->threadedCompositor();
```
```suggestion
    const auto* threadedCompositor = layerTreeHost->threadedCompositor();
```

Scony commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.h:
```diff
@@ -158,6 +160,9 @@ class ThreadedCompositor : public ThreadSafeRefCounted<ThreadedCompositor>, publ
     Ref<ThreadedDisplayRefreshMonitor> m_displayRefreshMonitor;
 #endif
+#if ENABLE(DAMAGE_TRACKING)
+    std::unique_ptr<WebCore::DamageForTesting> m_damageInfo;
```
that could be called `m_damageHistory`, `m_frameDamageHistory` or so

carlosgcampos commented on LayoutTests/TestExpectations:
```diff
@@ -7750,3 +7750,6 @@ webkit.org/b/286883 imported/w3c/web-platform-tests/css/CSS2/floats/remove-block
 [ Debug ] imported/w3c/web-platform-tests/css/CSS2/floats/line-pushed-by-floats-crash.html [ Skip ]
+
+# Damage-related tests are only supported on GTK and WPE
+damage/ [ Skip ]
```
Since damage are glib specific, I think they can be added to platform/glib/damage

spenap replied:
In the top most part of the expectations I mark them as skip, then in the glib expectations I mark them as pass. Is that not the way to do it?

carlosgcampos commented on Source/WebCore/page/ChromeClient.h:
```diff
@@ -728,6 +729,12 @@ class ChromeClient {
+#if ENABLE(DAMAGE_TRACKING)
+    virtual void resetDamageHistoryForTesting() { }
+
+    virtual WebCore::FrameDamageHistory* getDamageHistoryForTesting() const { return nullptr; }
```
```suggestion
    virtual WebCore::FrameDamageHistory* damageHistoryForTesting() const { return nullptr; }
```

carlosgcampos commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -7766,4 +7774,57 @@ void Internals::setShouldSkipResourceMonitorThrottling(bool flag)
+    if (!document || !document->page())
+        return Exception { ExceptionCode::NotSupportedError };
+
+    const auto& damageForTesting = document->page()->chrome().client().getDamageHistoryForTesting();
```
```suggestion
    auto* damageForTesting = document->page()->chrome().client().getDamageHistoryForTesting();
```

spenap replied:
used `const auto*`

carlosgcampos commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -7766,4 +7774,57 @@ void Internals::setShouldSkipResourceMonitorThrottling(bool flag)
+
+    Vector<Internals::FrameDamage> damageDetails;
+    const auto& damageData = damageForTesting->damageInformation();
+    for (size_t i = 0; i < damageData.size(); ++i) {
```
```suggestion
    for (auto& [isValid, region] : damageForTesting->damageInformation()) {
```

spenap replied:
used `const auto&` and defined a sequenceId outside of the loop as I used `i`

carlosgcampos commented on Source/WebCore/testing/Internals.h:
```diff
@@ -1546,6 +1550,18 @@ class Internals final
+#if ENABLE(DAMAGE_TRACKING)
+    using DamagePropagation = Damage::Propagation;
+    struct FrameDamage {
+        unsigned sequenceId;
```
```suggestion
        unsigned sequenceId { 0 };
```

carlosgcampos commented on Source/WebCore/testing/Internals.h:
```diff
@@ -1546,6 +1550,18 @@ class Internals final
+    using DamagePropagation = Damage::Propagation;
+    struct FrameDamage {
+        unsigned sequenceId;
+        bool isValid = false;
```
```suggestion
        bool isValid { false };
```

carlosgcampos commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -1994,4 +1999,23 @@ void WebChromeClient::didProgrammaticallyClearTextFormControl(const HTMLTextForm
+#if ENABLE(DAMAGE_TRACKING)
+void WebChromeClient::resetDamageHistoryForTesting()
+{
+    const auto* layerTreeHost = static_cast<LayerTreeHost*>(graphicsLayerFactory());
```
LayerTreeHost can be nullptr in gtk if accelerated compositing is disabled. I don't think we should be using LayerTreeHost from here, I would use the drawing area that is always available and that can access its layer tree host if present.

spenap replied:
thanks, I've done this through DrawingArea, as suggested, so we don't enable AC if layerTreeHost is not yet instantiated

carlosgcampos commented on LayoutTests/platform/glib/TestExpectations:
```diff
@@ -469,6 +469,9 @@ http/tests/websocket/tests/hybi/extensions.html [ Pass ]
 
+# Damage-related tests
+platform/glib/damage/ [ Pass ]
+
```
Now that these are glib specific tests we don't need this, they are expected to pass by default.

carlosgcampos reviewed (approved):
LGTM

spenap committed:
[GTK][WPE] Make it possible to verify damage rects in layout tests
https://bugs.webkit.org/show_bug.cgi?id=284951

Reviewed by Carlos Garcia Campos.

Expose damage information to JS through Internals for validation using
layout tests.

When DAMAGE_TRACKING is enabled and the Internals object is available,
frame damage information is stored and then made available through a new
FrameDamageForTesting interface. This is exposed through the
DrawingArea, and available in AC mode.

* LayoutTests/platform/glib/damage/basic-propagation-001-expected.txt: Added.
* LayoutTests/platform/glib/damage/basic-propagation-001.html: Added.
* LayoutTests/platform/glib/damage/basic-propagation-002-expected.txt: Added.
* LayoutTests/platform/glib/damage/basic-propagation-002.html: Added.
* LayoutTests/platform/glib/damage/common.css: Added.
(body):
* LayoutTests/platform/glib/damage/common.js: Added.
(failTest):
(setupTestCase):
* Source/WebCore/page/ChromeClient.h: New API to reset and get damage
  history.
(WebCore::ChromeClient::resetDamageHistoryForTesting):
(WebCore::ChromeClient::damageHistoryForTesting const):
* Source/WebCore/platform/graphics/Damage.h: New wrapper to store frame
  damage history.
(WebCore::FrameDamageHistory::damageInformation const):
(WebCore::FrameDamageHistory::addDamage):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::resetToConsistentState):
(WebCore::Internals::Internals):
(WebCore::Internals::getCurrentDamagePropagation const):
(WebCore::Internals::getFrameDamageHistory const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::resetDamageHistoryForTesting):
(WebKit::WebChromeClient::damageHistoryForTesting const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/DrawingAreaCoordinatedGraphics.cpp:
(WebKit::DrawingAreaCoordinatedGraphics::frameDamageForTesting const):
If AC mode is not enabled, frame damage history won't be available.
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/DrawingAreaCoordinatedGraphics.h:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/FrameDamageForTesting.h: Added.
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.h:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
(WebKit::ThreadedCompositor::paintToCurrentGLContext):
(WebKit::ThreadedCompositor::resetFrameDamageHistory):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.h:
(WebKit::ThreadedCompositor::frameDamageHistory const):
* Source/WebKit/WebProcess/WebPage/DrawingArea.h: Allow querying frame
  damage history for testing without side effects.
(WebKit::DrawingArea::frameDamageForTesting const):

Canonical link: https://commits.webkit.org/290948@main
